modules/nlu-testing/src/tools/visualisation.ts

```diff
@@ -0,0 +1,149 @@
+  for (const key in intentDatas) {
+    if (Object.prototype.hasOwnProperty.call(intentDatas, key)) {
+      const meanEmb = intentDatas[key]
+        .map(o => o.utt_emb)
+        .reduce((tot, cur) => {
+          return Matrix.add(tot, new Matrix([cur]))
+        }, Matrix.zeros(1, intentDatas[key][0].utt_emb.length))
+
+      intentEmb[key] = Matrix.div(meanEmb, intentDatas[key].length).to1DArray()
+    }
+  }
+
+  const indexAndIntent = Array.from(Object.keys(intentEmb).entries())
+  const simMat = { matrix: [], labels: [] }
+  for (const [index, key] of indexAndIntent) {
+    const row = []
+
+    for (const [indexTodo, keyTodo] of indexAndIntent) {
+      if (index === indexTodo) {
+        row.push(1)
+      } else {
+        row.push(_.round(similarity.cosine(intentEmb[key], intentEmb[keyTodo]), 2))
+      }
+    }
+
+    simMat.labels.push(key)
+    simMat.matrix.push(row)
+  }
+
+  const plotlyMatrixData = [
+    {
+      x: simMat.labels,
+      y: simMat.labels,
+      z: simMat.matrix,
+      type: 'heatmap'
+    }
+  ]
+
+  return plotlyMatrixData
+}
+
+export async function computeScatterEmbeddings(state: BotState, logger: sdk.Logger) {
+  const pcaTrain = new PCA(state.trainDatas.map(o => o.utt_emb))
+  const varianceTrain = pcaTrain.getExplainedVariance()
+
+  logger.info(
+    `Top 3 train variance ${varianceTrain.slice(0, 3).map(o => _.round(o, 2))} Accounting for ${_.round(
+      _.sum(varianceTrain.slice(0, 3)),
+      2
+    )}%`
+  )
+
+  const groupedIntentsTrain = _.groupBy(state.trainDatas, 'intent')
+  const traces = []
+
+  Object.entries(groupedIntentsTrain).map(([k, v]: [string, any[]], i) =>
+    traces.push({
+      x: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 0)),
+      y: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 1)),
+      z: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 2)),
+      mode: 'markers',
+      type: 'scatter3d',
+      name: k,
+      text: v.map(o => o.utt),
+      marker: { size: 8, color: i }
+    })
+  )
+
+  if (state.testDatas.length) {
+    const pcaTest = new PCA(state.testDatas.map(o => o.utt_emb))
+    const varianceTest = pcaTest.getExplainedVariance()
+
+    logger.info(
+      `Top 3 test variance ${varianceTest.slice(0, 3).map(o => _.round(o, 2))} Accounting for ${_.round(
+        _.sum(varianceTest.slice(0, 3)),
+        2
+      )}%`
+    )
+
+    const groupedIntentsTest = _.groupBy(state.testDatas, 'intent')
+    Object.entries(groupedIntentsTest).map(([k, v]: [string, any[]], i) =>
+      traces.push({
+        x: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 0)),
+        y: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 1)),
+        z: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 2)),
+        mode: 'markers',
+        type: 'scatter3d',
+        name: `${k}_test`,
+        text: v.map(o => o.utt),
+        marker: { size: 8, color: i }
+      })
+    )
+  }
+
+  return traces
+}
+
+function closest(a: number[], b: number[][], index: number): number {
```

Comment:
This does not compute the closest point of `b` form `a`... This finds the minimal distance. Rename this function for what it actually does.